src/execution_plan/ops/op_aggregate.c

```diff
@@ -6,127 +6,275 @@
 */
 
 #include "op_aggregate.h"
-#include "../../arithmetic/aggregate.h"
+#include "../../util/arr.h"
+#include "../../util/rmalloc.h"
 #include "../../grouping/group.h"
-#include "../../grouping/group_cache.h"
 #include "../../query_executor.h"
+#include "../../arithmetic/aggregate.h"
 
-OpBase* NewAggregateOp(AST *ast, TrieMap *groups) {
-    Aggregate *aggregate = malloc(sizeof(Aggregate));
-    aggregate->ast = ast;
-    aggregate->init = 0;
-    aggregate->none_aggregated_expression_count = 0;
-    aggregate->none_aggregated_expressions = NULL;
-    aggregate->group_keys = NULL;
-    aggregate->groups = groups;
+/* Construct an expression trees for both aggregated and none aggregated expressions. */
+static void _build_expressions(Aggregate *op) {
+    // RETURN expressions.
+    AST_ReturnNode *return_node = op->ast->returnNode;
+    uint expCount = array_len(return_node->returnElements);
+    
+    op->none_aggregated_expressions = array_new(AR_ExpNode*, 1);
+    op->expression_classification = rm_malloc(sizeof(int) * expCount);
 
-    OpBase_Init(&aggregate->op);
-    aggregate->op.name = "Aggregate";
-    aggregate->op.type = OPType_AGGREGATE;
-    aggregate->op.consume = AggregateConsume;
-    aggregate->op.reset = AggregateReset;
-    aggregate->op.free = AggregateFree;
+    for(uint i = 0; i < expCount; i++) {
+        AST_ReturnElementNode *returnElement = return_node->returnElements[i];
+        
+        AR_ExpNode *exp = AR_EXP_BuildFromAST(op->ast, returnElement->exp);
+        if(!AR_EXP_ContainsAggregation(exp, NULL)) {
+            op->expression_classification[i] = 0;
+            op->none_aggregated_expressions = array_append(op->none_aggregated_expressions, exp);
+        } else {
+            op->expression_classification[i] = 1;
+        }
+    }
 
-    return (OpBase*)aggregate;
+    // ORDER-BY expressions.
+    AST_OrderNode *order_node = op->ast->orderNode;
+    if(order_node) {
+        uint expCount = array_len(order_node->expressions);
```

Comment:
Shadows the `expCount` declared above; can just delete `uint`.

Reply:
👍